lib/puppet/util/package/version/rpm.rb

```diff
@@ -0,0 +1,287 @@
+            segment2 = ''
+          end
+        end
+
+        # if the segments we just grabbed from the strings are different types (i.e. one numeric one alpha),
+        # where alpha also includes ''; "numeric segments are always newer than alpha segments"
+        if segment2.empty?
+          return 1 if isnum
+
+          return -1
+        end
+
+        if isnum
+          # "throw away any leading zeros - it's a number, right?"
+          segment1 = segment1.gsub(/^0+/, '')
+          segment2 = segment2.gsub(/^0+/, '')
+          # "whichever number has more digits wins"
+          return 1 if segment1.length > segment2.length
+          return -1 if segment1.length < segment2.length
+        end
+
+        # "strcmp will return which one is greater - even if the two segments are alpha
+        # or if they are numeric. don't return if they are equal because there might
+        # be more segments to compare"
+        rc = segment1 <=> segment2
+        return rc if rc != 0
+      end # end while loop
+
+      # if we haven't returned anything yet, "whichever version still has characters left over wins"
+      if str1.length > str2.length
+        return 1
+      elsif str1.length < str2.length
+        return -1
+      else
+        return 0
+      end
+    end
+
+    # this method is a native implementation of the
+    # compare_values function in rpm's python bindings,
+    # found in python/header-py.c, as used by rpm.
+    def self.compare_values(s1, s2)
+      if s1.nil? && s2.nil?
+        return 0
+      elsif !s1.nil? && s2.nil?
+        return 1
+      elsif s1.nil? && !s2.nil?
+        return -1
+      end
+
+      rpmvercmp(s1, s2)
+    end
+
+    # how rpm compares two package versions:
+    # rpmUtils.miscutils.compareEVR(), which massages data types and then calls
+    # rpm.labelCompare(), found in rpm.git/python/header-py.c, which
+    # sets epoch to 0 if null, then compares epoch, then ver, then rel
+    # using compare_values() and returns the first non-0 result, else 0.
+    # This function combines the logic of compareEVR() and labelCompare().
+    #
+    # "yours" can be v, v-r, or e:v-r.
+    # "mine" will always be at least v-r, can be e:v-r
+    def self.rpm_compareEVR(yours, mine)
+      unless yours.epoch.nil?
+        rc = compare_values(yours.epoch, mine.epoch)
+        return rc unless rc == 0
+      end
+
+      rc = compare_values(yours.version, mine.version)
+      return rc unless rc == 0
+
+      # here is our special case, PUP-1244.
+      # if yours[:release] is nil (not specified by the user),
+      # and comparisons up to here are equal, return equal. We need to
+      # evaluate to whatever level of detail the user specified, so we
+      # don't end up upgrading or *downgrading* when not intended.
+      #
+      # This should NOT be triggered if we're trying to ensure latest.
+      return 0 if yours.release.nil?
+
+      rc = compare_values(yours.release, mine.release)
+
+      rc
```

Comment:
I don't think this line is needed as above line of code would cover the desired return value.